File: lintcode/combination-sum-ii.py
#!/usr/bin/env python
# encoding: utf-8

# Question: https://leetcode.com/problems/combination-sum-ii/

class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        stack = []
        res = []
        exist = set()
        for i in xrange(len(candidates)):
            if candidates[i] not in exist and target-candidates[i] >= 0:
                stack.append([[i], target-candidates[i]])
                exist.add(candidates[i])

        while stack:
            path, left = stack.pop()
            if left == 0:
                r = []
                for i in path:
                    r.append(candidates[i])
                res.append(r)
            elif left > 0:
                exist = set()
                for i in xrange(path[-1]+1, len(candidates)):
                    if candidates[i] not in exist and left >= candidates[i]:
                        path_copy = path[:]
                        path_copy.append(i)
                        stack.append([path_copy, left-candidates[i]])
                        exist.add(candidates[i])

        return res



if __name__ == '__main__':
    s = Solution()
    print s.combinationSum2([6,7,2,3], 7)
    print s.combinationSum2([10,1,2,7,6,1,5], 8)

File: leetcode-2/combination-sum-ii.cc
// Question: https://leetcode.com/problems/combination-sum-ii/

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> path;
        vector<vector<int>> ans;
        helper(candidates, 0, target, path, ans);
        return ans;
    }

    void helper(vector<int>& candidates, int start, int target, vector<int>& path, vector<vector<int>>& ans) {
        if (target == 0) {
            if (!path.empty()) {
                ans.push_back(path);
            }
        } else if (target > 0) {
            for (int i = start; i < candidates.size(); ++i) {
                if (i > start && candidates[i] == candidates[i-1]) {
                    continue;
                }
                path.push_back(candidates[i]);
                helper(candidates, i+1, target - candidates[i], path, ans);
                path.pop_back();
            }
        }
    }
};
